### PR TITLE
[POC] Implement callback for dask.

### PR DESCRIPTION
**Pull request by trivialfis**
Proof of concept implementation for dask training callback.  As it is not consistent with single node implementation, this is an early PR for review and discussion.

The rational behind a new style of callback function is that, the original one is modelled after R implementation.  The R impl uses language environment for storing states, while Python doesn't have such facility so the original callback creates an artificial one to be consistent with R.  I believe the new style is cleaner, native to Python, and easier customize.  With the new interface, https://github.com/dmlc/xgboost/pull/5238 and #5495 can be easily implemented.

To provide a short introduction to new callback function, here is its interface:
``` python
class Callback:
    def __init__(self):
        self.history = {}
        pass

    def before_training(self, model):
        pass

    def after_training(self, model):
        pass

    def before_iteration(self, model, epoch):
        return False

    def after_iteration(self, model, epoch):
        return False
```

All the other callbacks is sub-class of this `Callback`.  This interface is suitable for both single node and distributed training.  Function return value is whether training should be stopped,  `False` if not.  By using a return value, we can deprecate `EarlyStopException`, which is created for one particular callback.

To-dos:
- [ ] Make sure no data copy is performed.
- [ ] Get correct starting rounds: https://github.com/dmlc/xgboost/pull/5511 is closed so I need to figure out a way to get correct starting round.
- [x] Be compatible with original callback function.  The new `Callback` can be a wrapper of original callback, I need to port the implementation to non-dask training.
- [ ] Support user defined metric.
- [ ] Tests.

## Future plan

If the new design is approved, I would like to deprecate the following parameters:
- In scikit-learn interface `fit` function `eval_set`, `eval_metric`, `early_stopping_rounds`, `verbose`, `sample_weight_eval_set`, 
- From `train` function `evals`, `feval`, `maximize`, `early_stopping_rounds`, `evals_result`, `verbose_eval=True` 
- lastly corresponding parameters in dask.

## Related:

#4955
#3822

### CODE REVIEW

File: python-package/xgboost/dask.py [hunk 1]
```diff
@@ -365,14 +392,314 @@ def _get_rabit_args(worker_map, client):
     rabit_args = [('%s=%s' % item).encode() for item in env.items()]
     return rabit_args
 
+
+def evaluate(model, data, name):
```
Comment (trivialfis) [inline python-package/xgboost/dask.py:396]: Note to myself: Remove it, not used.

**Comment by trivialfis** [comment]
@hcho3 Can we disable the c++ header ordering lint?

**Comment by hcho3** [comment]
@trivialfis Why should we disable it? It is part of Google C++ coding style.
